plasma/limbo-priority-mod.md

```diff
@@ -0,0 +1,58 @@
+# Resolving Limbo Transactions via Exit Priority Modification
+
+Design out of conversation with vi, David Knott, Ben Jones, and Eva Beylin. 🍻
+
+---
+
+## TL;DR
+
+This post introduces an alternative to confirmation signatures in MVP that effectively allows any user to set the exit priority of a transaction (position in the priority queue when exiting). 
+
+## Background
+
+Plasma MVP relies on the concept of confirmation signatures to ensure that both parties to a transaction have data availability before it's considered complete. However, it appears that all proposed versions of the confirmation signature scheme are vulnerable to griefing and introduce annoying user experience artifacts. Generally, confirmation signatures operate on the assumption that all transactions are invalid until they are explicitly valid. 
+
+This has motivated the construction of alternative mechanisms which provide the same guarantees. We introduce a design whereby any party may attempt to set the exit priority of a transaction. This means that the transaction would now be processed earlier in the exit priority queue. This attempt is challengeable if the new priority would be greater than the priority of any of the transaction's inputs, if any of the inputs have been double-spent, or if the transaction's new priority would be lower than it's current priority. 
+
+The effect of this design is that an in-flight transaction that may or may not be included in the Plasma chain (a transaction "in limbo") can now be guaranteed to be included. Furthermore, if all of a transactions inputs were created at a time when the chain was valid, then the transaction will be successfully processed before any invalid transactions are processed.
+
+## Construction
+
+A user may attempt to set the priority of a transaction by submitting a signed transaction and a block number to the root chain, subject to a challenge period. This action is bonded, and the priority of deposit ("out of nowhere") transactions cannot be modified. If no successful challenge occurs during the challenge period, then the transaction will have a priority as if it were the added to the end of the given block. 
+
+A priority modification can be challenged in four ways (higher priority = would be exited sooner, lower priority = would be exited later):
+
+1. Someone shows that the specified transaction was included in a block, and that its original priority is higher than the new priority. This prevents someone from decreasing a transaction's priority. 
+2. Someone shows that any of the transaction's inputs would have a lower priority than the transaction. This prevents someone from moving a transaction "earlier" than any of the transaction's outputs.
+3. Someone requests that someone prove some input to the transaction has a higher priority than new priority. This is necessary if blocks are being withheld and a challenge of type (2) is not directly possible.
```

Comment:
Does scenario that requires challenge №3 to be used involves inputs that were not yet properly mined and revealed? 
`blknum` and `txindex` are not known to tx participants before block is mined and distributed. Does this mean that we need a change in tx format? I.e. addressing inputs by `txhash` instead of `blknum` and `txindex`?

Reply:
Yes. I think this requires (`txhash`, `oindex`).

Reply:
If we are not changing tx format, such tx chaining inside withheld blocks can be pulled off only by operator.

Reply:
My only worry is that we'd have to account for the new position whenever the priority changes, otherwise a transaction's inputs might be moved and then the transaction appears to be invalid (? maybe not as bad as I think if we can handle this dynamically). (txhash, oindex) doesn't have this problem and seemed simpler.